core/src/main/java/com/linecorp/armeria/client/HAProxyHandler.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static io.netty.handler.codec.haproxy.HAProxyCommand.PROXY;
+import static io.netty.handler.codec.haproxy.HAProxyProtocolVersion.V2;
+import static io.netty.handler.codec.haproxy.HAProxyProxiedProtocol.TCP4;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.proxy.HAProxyConfig;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.haproxy.HAProxyMessage;
+import io.netty.handler.codec.haproxy.HAProxyMessageEncoder;
+import io.netty.handler.proxy.ProxyConnectException;
+
+class HAProxyHandler extends ChannelOutboundHandlerAdapter {
+    private static final Logger logger = LoggerFactory.getLogger(HAProxyHandler.class);
+
+    private final HAProxyConfig haProxyConfig;
+
+    HAProxyHandler(HAProxyConfig haProxyConfig) {
+        this.haProxyConfig = haProxyConfig;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline().addBefore(ctx.name(), null, HAProxyMessageEncoder.INSTANCE);
+        super.handlerAdded(ctx);
+    }
+
+    @Override
+    public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress,
+                        SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        promise.addListener(f -> {
```

Comment:
I'm wondering if we need to complete this `promise` after `HAProxyMessage` is written successfully so that any message does not go through `HAProxyMessageEncoder` before it's removed. 🤔

Reply:
Even if `HAProxyMessageEncoder`remains in the pipeline, it is 1) stateless and 2) an `MessageToByteEncoder`, so I don't think it will really affect the actual http request.

Reply:
Thanks for the explanation. 😄